thunder/transforms/quantization.py

```diff
@@ -294,3 +295,214 @@ def transform_traces_pre_prologue(self, prologue_trace, computation_trace, epilo
 
         new_computation_trace.set_provenance(thunder.core.trace.TraceProvenance("quant pass"))
         return prologue_trace, new_computation_trace, epilogue_trace
+
+
+class LORATransform(Transform):
+    def __init__(
+        self,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        **kwargs,
+    ):
+        self.r = r
+        self.lora_alpha = lora_alpha
+        self.lora_dropout = lora_dropout
+        self.lora_linear_names = set()
+        self.lora_linear_states = {}
+
+    def transform_module(self, model: thunder.ThunderModule):
+        self.thunder_module = model
+        shared_names = model._get_shared_names()
+        processed_names = set()
+
+        def convert_linear_submodule(tm, name, r):
+            self.lora_linear_names.add(name)
+            weight_name = f"{name}.weight"
+            processed_copies = shared_names[weight_name] & processed_names
+            if processed_copies:
+                copy_name = next(iter(processed_copies))
+                self.lora_linear_states[weight_name] = self.lora_linear_states[copy_name]
+                tm._overrides_parameters[weight_name] = tm._overrides_parameters[copy_name]
+                tm._overrides_parameters[f"{weight_name}.loraA"] = tm._overrides_parameters[f"{copy_name}.loraA"]
+                tm._overrides_parameters[f"{weight_name}.loraB"] = tm._overrides_parameters[f"{copy_name}.loraB"]
+                processed_names.add(weight_name)
+                return
+
+            w = tm.get_parameter(weight_name)
+            in_features, out_features = w.shape[0], w.shape[1]
+
+            loraA = torch.nn.Parameter(torch.empty((r, in_features)))
+            torch.nn.init.kaiming_uniform_(loraA, a=math.sqrt(5))
+            loraB = torch.nn.Parameter(torch.empty((out_features, r)))
+            torch.nn.init.zeros_(loraB)
+
+            tm._overrides_parameters[weight_name] = w
+            tm._overrides_parameters[f"{weight_name}.loraA"] = loraA
+            tm._overrides_parameters[f"{weight_name}.loraB"] = loraB
+
+            self.lora_linear_states[weight_name] = {
+                "dtype": w.dtype,
+                "shape": tuple(w.shape),
+                "loraA.dtype": loraA.dtype,
+                "loraA.shape": loraA.shape,
+                "loraB.dtype": loraB.dtype,
+                "loraB.shape": loraB.shape,
+                "device": getattr(w, "_thunder_device", w.device),
+            }
+            processed_names.add(name)
+
+        for n, submodule in model._model.named_modules():
+            if isinstance(submodule, torch.nn.Linear):
+                convert_linear_submodule(model, n, self.r)
+
+    def transform_state_dict_for_submodule(
+        self, model: thunder.ThunderModule, submodule_name: str, state_dict: dict
+    ) -> dict:
+        if submodule_name not in self.lora_linear_names:
+            return state_dict
+
+        weight_name_full = f"{submodule_name}.weight"
```

Comment:
I don't think we need to do anything here with the weight, we just need to init the new loraA and loraB.
In particular w.to(w.device) is a no-op. (There is something funny in quantization because of the limitation described in  #1111 .)

Reply:
Does the original source code to subtract the lora initialization from the weight btw. (to leave the initial state unchanged in sum)? Then we should do this here and in transform module above.

Reply:
When checking [this](https://github.com/Lightning-AI/litgpt/blob/1d37f9a99bb4ba2b7373bc7fc5b8c5a457af48df/litgpt/lora.py#L564), I was not able to find where we subtract the initialization from the weight.

Reply:
Also, let me add lora_a and lora_b to state_dict!

Reply:
Yeah, with b initialized to zero, the contribution would initially be zero, so good.